File: BlinkCard/ios/blinkcard_flutter/Sources/blinkcard_flutter/BlinkCardArguments.swift
//
//  BlinkCardArguments.swift
//  blinkcard_flutter
//
//  Created by Milan Parađina on 26.02.2026..
//

import Foundation

enum BlinkCardPlatformMethods: String {
    case performScan
    case performDirectApiScan
    case loadSdk
    case unloadSdk
}

struct BlinkCardArguments {
    static let packageName = "blinkcard_flutter"
    static let blinkCardError = "blinkCardIos"
    static let sdkSettings = "blinkCardSdkSettings"
    static let sessionSettings = "blinkCardSessionSettings"
    static let scanningUxSettings = "scanningUxSettings"
}

enum BlinkCardFlutterError: Error, LocalizedError {
    case invalidArgumentsProvided
    case invalidLicenseKeyProvided
    case invalidSettingsProvided(message: String)
    case imageIsEmpty(message: String)
    case cancelled
    case initializationError(message: String)
    case error(message: String)
    
    var errorDescription: String? {
        switch self {
        case .invalidArgumentsProvided: return "Invalid method arguments provided!"
        case .invalidLicenseKeyProvided: return "Invalid license key provided"
        case .invalidSettingsProvided(let message): return "Invalid settings provided: \(message)"
        case .imageIsEmpty(let message): return message
        case .cancelled: return "Scanning has been cancelled"
        case .initializationError(let message): return "Initialization error: \(message)"
        case .error(let message): return "Error: \(message)"
        default: "Unknown error"
        }
    }
}

File: BlinkCard/android/src/main/kotlin/com/microblink/blinkcard/flutter/BlinkCardArguments.kt
package com.microblink.blinkcard.flutter

import com.microblink.blinkcard.flutter.BlinkCardPlatformMethods

enum class BlinkCardPlatformMethods(val method: String) {
    PERFORM_SCAN("performScan"),
    PERFORM_DIRECT_API_SCAN("performDirectApiScan"),
    LOAD_SDK("loadSdk"),
    UNLOAD_SDK("unloadSdk");

    companion object {
        fun fromMethod(method: String): BlinkCardPlatformMethods? {
            return values().firstOrNull {it.method == method }
        }
    }
}

object BlinkCardArguments {
    const val packageName: String = "blinkcard_flutter"
    const val blinkCardRequestCode: Int = 1453
    const val sdkSettings: String = "blinkCardSdkSettings"
    const val sessionSettings: String = "blinkCardSessionSettings"
    const val scanningUxSettings: String = "scanningUxSettings"
    const val blinkcardError = "blinkCardAndroid"
    const val firstSideImage = "firstSideImage"
    const val secondSideImage = "secondSideImage"
}

sealed class BlinkCardFlutterError(message: String): Exception(message) {

    class InvalidLicenseKeyProvided:
        BlinkCardFlutterError("Invalid license key provided")

    data class InvalidSettingsProvided(val detail: String) :
        BlinkCardFlutterError("Invalid settings provided: $detail")

    data class ImageIsEmpty(val detail: String):
        BlinkCardFlutterError(detail)

    class Cancelled :
        BlinkCardFlutterError("Scanning has been cancelled")

    data class InitializationError(val detail: String) :
        BlinkCardFlutterError("Initialization error: $detail")

    data class GenericError(val detail: String) :
        BlinkCardFlutterError("Error: $detail")
}